In [4]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sn
import numpy as np

In [5]:
#surpress warnings for clean notebook
import warnings
warnings.filterwarnings('ignore')

In [6]:
#read dataset
dataset = pd.read_csv("Downloads\machine learning practise\Melbourne_housing_FULL.csv")
dataset.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [7]:
dataset.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [8]:
dataset.shape

(34857, 21)

**columns that are usefull for our model**

In [9]:
col_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Distance', 'Bedroom2', 'Bathroom', 'Car', 'Landsize','BuildingArea',
             'CouncilArea', 'Regionname', 'Propertycount', 'Price']

In [10]:
dataset =  dataset[col_to_use]

In [11]:
dataset.head()

Suburb  Rooms Type Method SellerG  Distance  Bedroom2  Bathroom  Car  \
0  Abbotsford      2    h     SS  Jellis       2.5       2.0       1.0  1.0   
1  Abbotsford      2    h      S  Biggin       2.5       2.0       1.0  1.0   
2  Abbotsford      2    h      S  Biggin       2.5       2.0       1.0  0.0   
3  Abbotsford      3    u     VB  Rounds       2.5       3.0       2.0  1.0   
4  Abbotsford      3    h     SP  Biggin       2.5       3.0       2.0  0.0   

   Landsize  BuildingArea         CouncilArea             Regionname  \
0     126.0           NaN  Yarra City Council  Northern Metropolitan   
1     202.0           NaN  Yarra City Council  Northern Metropolitan   
2     156.0          79.0  Yarra City Council  Northern Metropolitan   
3       0.0           NaN  Yarra City Council  Northern Metropolitan   
4     134.0         150.0  Yarra City Council  Northern Metropolitan   

   Propertycount      Price  
0         4019.0        NaN  
1         4019.0  1480000.0  
2         4019.0  1035000.0  
3         4019.0        NaN  
4         4019.0  1465000.0

In [12]:
dataset.shape

(34857, 15)

In [13]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Distance             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
CouncilArea          3
Regionname           3
Propertycount        3
Price             7610
dtype: int64

**columns in which  NAN values are filled with zero**

In [14]:
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']

In [15]:
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

In [16]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Distance             0
Bedroom2             0
Bathroom             0
Car                  0
Landsize         11810
BuildingArea     21115
CouncilArea          3
Regionname           3
Propertycount        0
Price             7610
dtype: int64

**columns in which NAN values are filled with their mean values**

In [17]:
dataset['Landsize'] = dataset['Landsize'].fillna(dataset['Landsize'].mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset['BuildingArea'].mean())

In [31]:
dataset.isna().sum()

Suburb              0
Rooms               0
Type                0
Method              0
SellerG             0
Distance            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
CouncilArea         3
Regionname          3
Propertycount       0
Price            7610
dtype: int64

**rows which consists NAN values are dropped**

In [18]:
dataset.dropna(inplace = True)

In [19]:
dataset.isna().sum()

Suburb           0
Rooms            0
Type             0
Method           0
SellerG          0
Distance         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
CouncilArea      0
Regionname       0
Propertycount    0
Price            0
dtype: int64

In [20]:
dataset.shape

(27244, 15)

**get dummy variables for category type variables and first columns of them are dropped to avoid dummy variable trap**

In [21]:
dataset = pd.get_dummies(dataset, drop_first = True)

In [22]:
dataset.head()

Rooms  Distance  Bedroom2  Bathroom  Car  Landsize  BuildingArea  \
1      2       2.5       2.0       1.0  1.0     202.0      160.2564   
2      2       2.5       2.0       1.0  0.0     156.0       79.0000   
4      3       2.5       3.0       2.0  0.0     134.0      150.0000   
5      3       2.5       3.0       2.0  1.0      94.0      160.2564   
6      4       2.5       3.0       1.0  2.0     120.0      142.0000   

   Propertycount      Price  Suburb_Aberfeldie  ...  \
1         4019.0  1480000.0                  0  ...   
2         4019.0  1035000.0                  0  ...   
4         4019.0  1465000.0                  0  ...   
5         4019.0   850000.0                  0  ...   
6         4019.0  1600000.0                  0  ...   

   CouncilArea_Wyndham City Council  CouncilArea_Yarra City Council  \
1                                 0                               1   
2                                 0                               1   
4                                 0                               1   
5                                 0                               1   
6                                 0                               1   

   CouncilArea_Yarra Ranges Shire Council  Regionname_Eastern Victoria  \
1                                       0                            0   
2                                       0                            0   
4                                       0                            0   
5                                       0                            0   
6                                       0                            0   

   Regionname_Northern Metropolitan  Regionname_Northern Victoria  \
1                                 1                             0   
2                                 1                             0   
4                                 1                             0   
5                                 1                             0   
6                                 1                             0   

   Regionname_South-Eastern Metropolitan  Regionname_Southern Metropolitan  \
1                                      0                                 0   
2                                      0                                 0   
4                                      0                                 0   
5                                      0                                 0   
6                                      0                                 0   

   Regionname_Western Metropolitan  Regionname_Western Victoria  
1                                0                            0  
2                                0                            0  
4                                0                            0  
5                                0                            0  
6                                0                            0  

[5 rows x 745 columns]

**define independent variables and dependent variable**

In [23]:
X = dataset.drop('Price', axis = 'columns')
y = dataset['Price']

**split data into train and test set**

In [24]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

**Start building Linear Regression Model**

In [25]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [26]:
model.fit(X_train, y_train)

LinearRegression()

In [27]:
model.score(X_train, y_train)

0.6890660980148369

In [28]:
model.score(X_test, y_test)

0.6400936043606525

In [29]:
from sklearn import linear_model

**Linear Regression with L1 regularization**

In [30]:
lasso_model = linear_model.Lasso(alpha = 50, max_iter = 100, tol =  0.1)

In [31]:
lasso_model.fit(X_train, y_train)

Lasso(alpha=50, max_iter=100, tol=0.1)

In [32]:
lasso_model.score(X_train, y_train)

0.6848048692208911

In [33]:
lasso_model.score(X_test, y_test)

0.641972801987357

**Linear Regression  with L2 regularization**

In [34]:
ridge_model = linear_model.Ridge(alpha = 50, max_iter = 100, tol =  0.1)

In [35]:
ridge_model.fit(X_train, y_train)

Ridge(alpha=50, max_iter=100, tol=0.1)

In [36]:
ridge_model.score(X_train, y_train)

0.6719111433510354

In [37]:
ridge_model.score(X_test, y_test)

0.6336020978156653

**we are getting accuracy around 64% in each model(LinearRegression, Regression with L1 regularization , Regression with L2)**